### PR TITLE
Redis caching for redirects

### PR DESCRIPTION
**Pull request by rogierslag**
This adds additional caching, making eg `nginx` based caching no longer required

### CODE REVIEW

File: src/caches/redis.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+  async function getFromCache(params) {
+    const metric = timingMetric(CACHE, {tags: {cacheOperation: 'getFromCache'}});
+    const result = await get(key(params));
+    if (result) {
+      // Cache hit
+      metrics.write(metric);
+    } else {
+      // Cache miss
+      metrics.write(metric);
```
Comment (AlexKolpa) [inline src/caches/redis.js]: Pretty sure both branches do the same?

File: src/caches/redis.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+const DURATION_1_DAY = 24 * 60 * 60;
+
+function key(params) {
+  return Object.entries(params).map(([key, value]) => `${key}=${value}`).join('-');
+}
+
+export default function redis() {
+  const url = config.get('redis.url');
+
+  const client = redisClient.createClient({url, prefix: 'iaas--'});
```
Comment (AlexKolpa) [inline src/caches/redis.js:17]: Just curious if the triple dashes separator is based on something or if it's used elsewhere?

File: src/caches/redis.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  async function addToCache(params, url) {
+    const metric = timingMetric(CACHE, {tags: {cacheOperation: 'addToCache'}});
+    try {
+      await setex(key(params), DURATION_1_DAY, url);
+      return true;
+    } catch (e) {
+      const message = e.toString();
+      log('error', message);
```
Comment (AlexKolpa) [inline src/caches/redis.js]: Maybe include a descriptive message such as `error adding ${url} for ${key} to cache: ${message}`?

File: src/fastCache.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import log from "./log";
+import cache from "./caches";
+
+export async function getFromCache(params) {
```
Comment (AlexKolpa) [inline src/fastCache.js:4]: Seeing as you do the `* as fastCache` import below already as well, the `from/toCache` feel a bit double.

File: src/imageResponse.js [hunk 5]
```diff
@@ -167,6 +168,19 @@ export async function magic(params, method, response, stats = undefined, metric
     }
     log('debug', `Request for ${imageDescription}`);
 
+    const fastCacheValue = await fastCache.getFromCache(params);
+    if (fastCacheValue) {
+      log('debug', `Fast cache hit for ${imageDescription}`);
+      redirectToCachedEntity(fastCacheValue, params, response);
+      if (metric) {
```
Comment (AlexKolpa) [inline src/imageResponse.js:175]: Maybe include a flag that the value was a cache hit/miss?

File: .nvmrc [hunk 6]
```diff
@@ -1 +1 @@
-8.15
+10.16
```
Comment (AlexKolpa) [inline .nvmrc:1]: No major issues with this upgrade I presume?

File: Dockerfile [hunk 7]
```diff
@@ -31,12 +31,14 @@ RUN cd /opt/iaas && yarn install --pure-lockfile
 ADD src/*.js /opt/iaas/src/
```
Comment (AlexKolpa) [inline Dockerfile:31]: Only noticed this now, but is there a specific reason you're not just doing `COPY src /opt/iaas/src` and then the babel compilation in one go for all of `src`?

File: src/fastCache.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import log from "./log";
```
Comment (AlexKolpa) [inline src/fastCache.js:1]: The naming confuses me. You got `caches` which wraps `redis` (and presumably could support other cache sources such as local memory in the future?), but then you have this somewhat out of place `fastCache`? Can't you merge this logic into `caches`?

**Review by AlexKolpa — APPROVED** [review]
Most of it looks good, just a few questions

File: .nvmrc (hunk 6 above)
Comment (rogierslag) [reply .nvmrc:1]: Docker container already ran with this. Apparently forgot to up it locally

File: src/caches/redis.js (hunk 2 above)
Comment (rogierslag) [reply src/caches/redis.js:17]: just to put more distance between the key and its "values"

File: src/fastCache.js (hunk 8 above)
Comment (rogierslag) [reply src/fastCache.js:1]: the `fastCache` is the abstraction if no cache has been configured

**Pull request merged**